File: assets/lib/boilerplate.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Windows.Forms;
using OpenBveApi.Runtime;

// Start of BlocklyAts boilerplate code.
internal class BlocklyAtsCompanion {
    
    public IRuntime Plugin;
  
    public VehicleSpecs VSpec;
    public ElapseData EData;
    public int[] Panel = new int[256];
    public DoorStates DoorState;
    public bool LegacyDoorState {
        get {
            return this.DoorState != DoorStates.None;
        }
    }
    public bool[] KeyState = new bool[16];
    
    public class TimerTuple {
        public double Interval;
        public double LastTrigger;
        public bool Cycle;
        public bool Enabled;
    }
    
    public Dictionary<string, TimerTuple> TimerState = new Dictionary<string, TimerTuple>();
    public Dictionary<string, Dictionary<string, dynamic>> ConfigData;
    
    private string pluginDir = Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location).TrimEnd('\\', '/');
    private PlaySoundDelegate playSoundDelegate;
    private SoundHandle[] soundHandles = new SoundHandle[256];
    // To achieve the same behavior as Win32 plugin interface at bve_get_sound_internal
    private int[] emulatedSoundState = new int[256];
    
    public BlocklyAtsCompanion(LoadProperties prop, IRuntime plugin) {
        playSoundDelegate = prop.PlaySound;
        prop.Panel = Panel;
        for (int i = 0; i < 256; i++) emulatedSoundState[i] = -10000;
        this.Plugin = plugin;
    }
  
    public void LoadConfig(string path) {
      if (!File.Exists(Path.Combine(pluginDir, path))) return;
      ConfigData = new Dictionary<string, Dictionary<string, dynamic>>();
      var section = "";
      foreach (var line in System.IO.File.ReadAllLines(Path.Combine(pluginDir, path))) {
          var tline = line.Trim();
          if (tline.StartsWith("[") && tline.EndsWith("]")) {
              section = tline.Substring(1, tline.Length - 2).Trim().ToLowerInvariant();
              if (!ConfigData.ContainsKey(section)) ConfigData[section] = new Dictionary<string, dynamic>();
          }
          if (!tline.StartsWith("#") && tline.Contains("=")) {
              var parts = tline.Split(new[] { '=' }, 2);
              string confKey = parts[0].Trim().ToLowerInvariant();
              string confValue = parts[1].Trim();
              double td;
              if (double.TryParse(confValue, out td)) ConfigData[section][confKey] = td;
              else if (confValue.ToLowerInvariant() == "true") ConfigData[section][confKey] = true;
              else if (confValue.ToLowerInvariant() == "false") ConfigData[section][confKey] = false;
              else ConfigData[section][confKey] = confValue;
          }
      }
    }

    public void SaveConfig(string path) {
      var sb = new System.Text.StringBuilder();
      foreach (var section in ConfigData) {
          sb.AppendFormat("[{0}]\n", section.Key);
          foreach (var pair in section.Value) {
              sb.AppendFormat("{0}={1}\n", pair.Key, pair.Value);
          }
          sb.AppendLine();
      }
      System.IO.File.WriteAllText(Path.Combine(pluginDir, path), sb.ToString());
    }

    public int AccessLegacySound(int id, int state = int.MaxValue, double volume = double.NaN) {
      if (state == 0 && double.IsNaN(volume)) {
          if (volume <= 0) {
              state = -10000;
          } else if (volume >= 100) {
              state = 0;
          } else {
              state = (int)(volume * 100) - 10000;
          }
      }
      if (state == -10000) {
          if (soundHandles[id] != null && soundHandles[id].Playing) {
              soundHandles[id].Stop();
              soundHandles[id] = null;
          }
          emulatedSoundState[id] = state;
      } else if (state > -10000 && state <= 0) {
          if (soundHandles[id] != null && soundHandles[id].Playing) {
              soundHandles[id].Volume = (state + 10000) / 10000;
          } else {
              soundHandles[id] = playSoundDelegate(id, (state + 10000) / 10000, 1, true);
          }
          emulatedSoundState[id] = state;
      } else if (state == 1) {
          if (soundHandles[id] != null && soundHandles[id].Playing) {
              soundHandles[id].Stop();
          }
          soundHandles[id] = playSoundDelegate(id, 1, 1, false);
          emulatedSoundState[id] = 2;
      } else if (state == 2) {
          emulatedSoundState[id] = state;
      } else if (state > 10000) {
          return emulatedSoundState[id];
      } else {
          throw new ArgumentOutOfRangeException();
      }
      return 0;
    }
    
    public dynamic GetConfig(string part, string key, dynamic defaultValue = null) {
        part = part.Trim().ToLowerInvariant();
        key = key.Trim().ToLowerInvariant();
        if (ConfigData != null && ConfigData.ContainsKey(part) && ConfigData[part].ContainsKey(key)) {
            return ConfigData[part][key];
        } else {
            return defaultValue;
        }
    }
    
    public void SetConfig(string part, string key, dynamic confValue = null) {
        part = part.Trim().ToLowerInvariant();
        key = key.Trim().ToLowerInvariant();
        if (ConfigData == null) ConfigData = new Dictionary<string, Dictionary<string, dynamic>>();
        if (!ConfigData.ContainsKey(part)) ConfigData.Add(part, new Dictionary<string, dynamic>());
        if (!ConfigData[part].ContainsKey(key)) {
            ConfigData[part].Add(key, confValue);
        } else {
            ConfigData[part][key] = confValue;
        }
    }
    
    private void CallTimerHandler(string name) {
        var methodInfo = Plugin.GetType().GetMethod("_etimertick_" + name);
        if (methodInfo != null) methodInfo.Invoke(Plugin, new object[] { });
    }
    
    public void ResetTimer(string name, bool callHandler) {
        if (!TimerState.ContainsKey(name)) return;
        if (EData == null) {
            // Force a timer reset later if the ElapseData is currently not ready.
            TimerState[name].LastTrigger = double.MaxValue;
        } else {
            TimerState[name].LastTrigger = EData.TotalTime.Milliseconds;
        }
        TimerState[name].Enabled = true;
        if (callHandler) CallTimerHandler(name);
    }
    
    public void CancelTimer(string name, bool callHandler) {
        if (!TimerState.ContainsKey(name)) return;
        TimerState[name].Enabled = false;
        if (callHandler) CallTimerHandler(name);
    }
    
    public void SetTimer(string name, double interval, bool cycle) {
        if (interval > 0) {
            TimerState[name] = new TimerTuple();
            TimerState[name].Interval = interval;
            TimerState[name].Cycle = cycle;
            ResetTimer(name, false);
        } else {
            CancelTimer(name, false);
        }
    }
    
    public void UpdateTimer() {
        foreach (var entry in TimerState) {
            if (!entry.Value.Enabled) continue;
            if (entry.Value.LastTrigger > EData.TotalTime.Milliseconds) {
                // From the future? Maybe caused by a station jump, just reset it
                ResetTimer(entry.Key, false);
            } else if (EData.TotalTime.Milliseconds >= entry.Value.LastTrigger + entry.Value.Interval) {
                CallTimerHandler(entry.Key);
                if (entry.Value.Cycle) {
                    ResetTimer(entry.Key, false);
                } else {
                    CancelTimer(entry.Key, false);
                }
            }
        }
    }
}

File: assets/lib/boilerplate.lua
require "winfile"

-- Start of BlocklyAts boilerplate code.
local LIP={}function LIP.load(b)assert(type(b)=='string','Parameter "fileName" must be a string.')local c=winfile.open(b,'r');if c==nil then;return;end;local d={}local e;for f in c:lines()do local g=f:match('^%[([^%[%]]+)%]$')if g then e=tonumber(g)and tonumber(g)or g;d[e]=d[e]or{}end;local h,i=f:match('^([%w|_]+)%s-=%s-(.+)$')if h and i~=nil then if tonumber(i)then i=tonumber(i)elseif i=='true'then i=true elseif i=='false'then i=false end;if tonumber(h)then h=tonumber(h)end;d[e][h]=i end end;c:close()return d end;function LIP.save(b,d)assert(type(b)=='string','Parameter "fileName" must be a string.')assert(type(d)=='table','Parameter "data" must be a table.')local c=assert(winfile.open(b,'w+b'),'Error loading file :'..b)local j=''for e,h in pairs(d)do j=j..('[%s]\n'):format(e)for k,i in pairs(h)do j=j..('%s=%s\n'):format(k,tostring(i))end;j=j..'\n'end;c:write(j)c:close()end

_bkeystate = {false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false}
_bdoorstate = false
_vtimer = {}

function _fcfgload(path)
  _vconfig = LIP.load(_vdlldir..path)
end

function _fcfgsave(path)
  LIP.save(_vdlldir..path, _vconfig)
end

function _fcfgget(part, key, default)
  if _vconfig == nil then; return default; end
  if _vconfig[part] == nil then; return default; end
  if _vconfig[part][key] == nil then; return default; end
  return _vconfig[part][key]
end

function _fcfgset(part, key, value)
  if _vconfig == nil then; _vconfig = {}; end
  if _vconfig[part] == nil then; _vconfig[part] = {}; end
  _vconfig[part][key] = value
end

-- Timer functions
-- _vtimer[<NAME>][1]:  Interval
-- _vtimer[<NAME>][2]:  Should timer repeat?
-- _vtimer[<NAME>][3]:  The time of last trigger
-- _vtimer[<NAME>][4]:  Is enabled?

function _ftimercallhandler(name)
  if _G["_etimertick_"..name] ~= nil then; _G["_etimertick_"..name](); end
end

function _ftimerreset(name, callhandler)
  if _vtimer[name] == nil then; return; end
  _vtimer[name][3] = _bEdTime
  _vtimer[name][4] = true
  if callhandler then; _ftimercallhandler(name); end
end

function _ftimercancel(name, callhandler)
  if _vtimer[name] == nil then; return; end
  _vtimer[name][4] = false;
  if callhandler then; _ftimercallhandler(name); end
end

function _ftimerset(name, interval, cycle)
  if interval > 0 then
    _vtimer[name] = { interval, cycle, 0, false }
    _ftimerreset(name, false)
  else
    _ftimercancel(name, false)
  end
end

function _ftimerupdate()
  for name, param in pairs(_vtimer) do
    if param[4] then
      if param[3] == nil or param[3] > _bEdTime then
        -- Nil means the timer was initialized when ElapseData was not ready
        -- Or from the future? Maybe caused by a station jump, just reset it
        _ftimerreset(name, false)
      elseif _bEdTime >= param[3] + param[1] then
          _ftimercallhandler(name)
          if param[2] then; _ftimerreset(name, false); else; _ftimercancel(name, false); end
      end
    end
  end
end